File: 72/edit-distance.cpp
class Solution {
public:
    int minDistance(string word1, string word2) {
        int N1 = word1.length(), N2 = word2.length();
        int table[N1+1][N2+1];
        for(int i=0; i<=N1; i++) table[i][0] = i; //"" 走到 word1[i] 距離是i
        for(int j=0; j<=N2; j++) table[0][j] = j; //"" 走到 word2[j] 距離是j

        for(int i=1; i<=N1; i++) {
            for(int j=1; j<=N2; j++) {
                if(word1[i-1]==word2[j-1]) {
                    table[i][j] = table[i-1][j-1]; //相同不用編輯，edit距離沒增加
                } else {
                    int d1 = table[i-1][j-1] + 1; //修改取代1個字母
                    int d2 = table[i-1][j] + 1; //刪除1個字母
                    int d3 = table[i][j-1] + 1; //增加1個字母
                    table[i][j] = min(d1,d2,d3);
                }
            }
        }
        return table[N1][N2];
    }
    int min(int a, int b, int c) {
        if(a<=b && a<=c) return a;
        if(b<=a && b<=c) return b;
        return c;
    }
};

File: 72/edit-distance.py
class Solution:
    def minDistance(self, word1: str, word2: str) -> int:
        N1 = len(word1) + 1
        N2 = len(word2) + 1
        # table = [[0]*N2]*N1 # 這種寫法是錯的，因為變成都指到同一個
        # 下面寫法才是對的，不過這樣看起來 Python 好像也沒多簡潔...
        table = [[0 for _ in range(N2)] for _ in range(N1)]
        for i in range(N1):
            table[i][0] = i
        for j in range(N2):
            table[0][j] = j
        
        for i in range(1,N1):
            for j in range(1,N2):
                if word1[i-1] == word2[j-1]:
                    table[i][j] = table[i-1][j-1]
                else:
                    d1 = table[i-1][j] + 1
                    d2 = table[i][j-1] + 1
                    d3 = table[i-1][j-1] + 1
                    table[i][j] = min(d1,d2,d3)
        return table[N1-1][N2-1]
